src/base/castlelocalisation.pas

```diff
@@ -0,0 +1,223 @@
+  FOnLocalisationUpdatedEventList := TOnLocalisationUpdatedEventList.Create;
+  FOnUpdateLocalisationEventList := TOnUpdateLocalisationEventList.Create;
+end;
+
+destructor TCastleLocalisation.Destroy;
+begin
+  FreeAndNil(FOnUpdateLocalisationEventList);
+  FreeAndNil(FOnLocalisationUpdatedEventList);
+  FreeAndNil(FLocalisationIDList);
+  FreeAndNil(FFileLoaderDictionary);
+  FreeAndNil(FLanguageDictionary);
+
+  inherited;
+end;
+
+/////////////////////
+//Private/Protected//
+/////////////////////
+
+function TCastleLocalisation.Get(AKey: String): String;
+begin
+  if not FLanguageDictionary.TryGetValue(AKey, Result) then
+    Result := AKey; //When no translation is found, return the key.
+end;
+
+procedure TCastleLocalisation.LoadLanguage(const ALanguageURL: String);
+var
+  FileLoaderAction: TFileLoaderAction;
+  LocalisedText: String;
+  OnUpdateLocalisationEvent: TOnUpdateLocalisationEvent;
+begin
+  if FLanguageURL = ALanguageURL then Exit;
+  FLanguageURL := ALanguageURL;
+
+  FLanguageDictionary.Clear;
+
+  if ALanguageURL = '' then Exit; //If there's no language XML file, then that's it, no more localisation.
+
+  FFileLoaderDictionary.TryGetValue(ExtractFileExt(ALanguageURL), FileLoaderAction);
+  Check(Assigned(FileLoaderAction), 'There is no file loader associated with the extension of the given file.');
+
+  FileLoaderAction(AbsoluteURI(ALanguageURL), FLanguageDictionary);
+
+  //Tell every registered object to update its localisation:
+  for OnUpdateLocalisationEvent in FOnUpdateLocalisationEventList do
+  begin
+    FLocalisationIDList.TryGetValue(OnUpdateLocalisationEvent, LocalisedText);
+    OnUpdateLocalisationEvent(Items[LocalisedText]);
+  end;
+end;
+
+procedure TCastleLocalisation.Notification(AComponent: TComponent; Operation: TOperation);
+var
+  LCastleLocalisationComponent: ICastleLocalisation;
+begin
+  if Operation = opRemove then
+  begin
+    AComponent.RemoveFreeNotification(Self);
+
+    LCastleLocalisationComponent := AComponent as ICastleLocalisation;
+    FOnUpdateLocalisationEventList.Remove(@LCastleLocalisationComponent.OnUpdateLocalisation);
+    FLocalisationIDList.Remove(@LCastleLocalisationComponent.OnUpdateLocalisation);
+  end;
+end;
+
+//////////////
+////Public////
+//////////////
+
+function TCastleLocalisation.SystemLanguage(const ADefaultLanguage: String = SystemDefaultLanguage): String;
+begin
+  Result := CastleSystemLanguage.SystemLanguage(ADefaultLanguage);
+end;
+
+function TCastleLocalisation.SystemLocal(const ADefaultLocal: String = SystemDefaultLocal): String;
+begin
+  Result := CastleSystemLanguage.SystemLocal(ADefaultLocal);
+end;
+
+procedure TCastleLocalisation.AddOrSet(ALocalisationComponent: ICastleLocalisation; ALocalisationID: String);
+var
+  IsNewEntry: Boolean;
+begin
+  if ALocalisationID = '' then
+    Exit;
+
+  IsNewEntry := true;
+  try
+    FLocalisationIDList.Add(@ALocalisationComponent.OnUpdateLocalisation, ALocalisationID);
+  except
```

Comment:
If possible, do not react to all possible exceptions.

1. better to catch only the exception classes that mean "the item was already existing". Looking at Generics.Collections sources, it seems you want to capture `EListError`.

2. Alternative is to do it like:

```
IsNewEntry := FLocalisationIDList.ContainsKey(@ALocalisationComponent.OnUpdateLocalisation);
FLocalisationIDList.AddOrSetValue(@ALocalisationComponent.OnUpdateLocalisation, ALocalisationID);
```

It's a *bit* slower (you're doing a `ContainsKey` check that will be essentially repeated inside `AddOrSetValue`), but it looks simple.

Both options 1. and 2. are OK by me:)